background.js

```diff
@@ -10,21 +21,159 @@ chrome.runtime.onMessage.addListener(async (message) => {
+      'match_urls',
+      'exclude_urls',
+    ]);
+
+    console.debug(`[Auto-Archive Debug] Settings: enable_auto_archive=${enable_auto_archive}, match_urls="${match_urls}", exclude_urls="${exclude_urls}"`);
+
+    if (!enable_auto_archive || !match_urls || match_urls.trim() === '') {
+      console.debug('[Auto-Archive Debug] Auto-archiving disabled or match pattern empty');
+      return false;
+    }
+
+    const matchPattern = new RegExp(match_urls);
+    const matches = matchPattern.test(url);
+    console.debug(`[Auto-Archive Debug] URL match test: ${matches} (pattern: ${matchPattern})`);
+
+    if (!matches) {
+      return false;
+    }
+
+    if (exclude_urls.trim()) {
+      try {
+        const excludePattern = new RegExp(exclude_urls);
+        const excluded = excludePattern.test(url);
+        console.debug(`[Auto-Archive Debug] URL exclude test: ${excluded} (pattern: ${excludePattern})`);
+
+        if (excluded) {
+          return false;
+        }
+      } catch (error) {
+        console.error('Invalid exclude pattern:', error);
+      }
+    }
+
+    console.debug(`[Auto-Archive Debug] URL ${url} should be archived: TRUE`);
+    return true;
+  } catch (error) {
+    console.error('Error checking auto-archive patterns:', error);
+    return false;
+  }
+}
+
+// Global reference to the listener so we can remove it properly
+let tabUpdateListener = null;
+
+async function setupAutoArchiving() {
+  console.debug('[Auto-Archive Debug] Setting up auto-archiving...');
+
+  const hasPermission = await chrome.permissions.contains({ permissions: ['tabs'] });
+  console.debug(`[Auto-Archive Debug] Has tabs permission: ${hasPermission}`);
+
+  if (!hasPermission) {
+    console.log('Tabs permission not granted, auto-archiving disabled');
+    // Reset the toggle if permission was not granted
+    chrome.storage.local.set({ enable_auto_archive: false });
+    return;
+  }
+
+  if (tabUpdateListener) {
+    try {
+      chrome.tabs.onUpdated.removeListener(tabUpdateListener);
+      tabUpdateListener = null;
+    } catch (error) {
+      // ignore
+    }
+  }
+
+  const { enable_auto_archive=false } = await chrome.storage.local.get(['enable_auto_archive']);
+  console.debug(`[Auto-Archive Debug] enable_auto_archive setting: ${enable_auto_archive}`);
+
+  if (enable_auto_archive) {
+    tabUpdateListener = async (tabId, changeInfo, tab) => {
+      console.debug(`[Auto-Archive Debug] Tab updated - tabId: ${tabId}, status: ${changeInfo.status}, url: ${tab?.url}`);
+
+      // Only process when the page has completed loading
+      if (changeInfo.status === 'complete' && tab.url) {
+        console.debug(`[Auto-Archive Debug] Tab load complete, checking if URL should be archived: ${tab.url}`);
+
+        // Check if URL is already archived locally
+        const { snapshots = [] } = await chrome.storage.local.get('snapshots');
+        const isAlreadyArchived = snapshots.some(s => s.url.trim() === tab.url.trim());
```

Comment:
The trim here is unnecessary, URLs should never have leading or trailing whitespace
```suggestion
        const isAlreadyArchived = snapshots.some(s => s.url === tab.url);
```